spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java

```diff
@@ -272,9 +274,40 @@ private Entity convertToEntityForSave(Object entity, Key... ancestors) {
 		Builder builder = Entity.newBuilder(key);
 		this.datastoreEntityConverter.write(entity, builder);
 		saveDescendents(entity, key);
+		saveReferences(entity, builder);
 		return builder.build();
 	}
 
+	private void saveReferences(Object entity, Builder builder) {
+		DatastorePersistentEntity datastorePersistentEntity = this.datastoreMappingContext
+				.getPersistentEntity(entity.getClass());
+		datastorePersistentEntity.doWithReferenceProperties(persistentProperty -> {
+			PersistentPropertyAccessor accessor = datastorePersistentEntity.getPropertyAccessor(entity);
+			Object val = accessor.getProperty(persistentProperty);
+			if (val == null) {
+				return;
+			}
+			Value<?> value;
+			if (persistentProperty.isCollectionLike()) {
+				if (val.getClass().isArray() && val.getClass() != byte[].class) {
+					//if a property is an array, convert it to list
+					val = CollectionUtils.arrayToList(val);
+				}
+				saveAll((Iterable<?>) val);
+				List<KeyValue> keyValues = StreamSupport.stream(((Iterable<?>) val).spliterator(), false)
+						.map(o -> KeyValue.of(this.getKey(o, false)))
```

Comment:
Here we're not allocating new keys, but should we? I'm not thinking of any downsides right now to just allocating the keys for them, did you have some in mind?

Reply:
saveAll takes care of creating keys, so we don't need it to allocate them here